packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { CompletionItemKind, SnippetString } from 'vscode';
+import ProtocolCompletionItem from 'vscode-languageclient/lib/protocolCompletionItem';
+import { retrieveSObject, retrieveSObjects } from '../sfdx';
+
+import { Middleware } from 'vscode-languageclient';
+
+const EXPANDABLE_ITEM_PATTERN = /__([A-Z_]+)/;
+
+export const middleware: Middleware = {
+  // The SOQL LSP server may include special completion items as "placeholders" for
+  // the client to expand with information from the users' default Salesforce Org.
+  // We do that here as middleware, transforming the server response before passing
+  // it up to VSCode.
+
+  provideCompletionItem: async (document, position, context, token, next) => {
+    const items = (await next(
+      document,
+      position,
+      context,
+      token
+    )) as ProtocolCompletionItem[];
+
+    return expandPlaceholders(await filterByContext(items));
+  }
+};
+
+interface SoqlItemContext {
+  sobjectName: string;
+  fieldName?: string;
+  notNillable?: boolean;
+}
+
+async function filterByContext(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const filteredItems: ProtocolCompletionItem[] = [];
+
+  for (const item of items) {
+    if (
+      !EXPANDABLE_ITEM_PATTERN.test(item.label) &&
+      item?.data?.soqlContext?.sobjectName
+    ) {
+      const objMetadata = await retrieveSObject(
+        item.data.soqlContext.sobjectName
+      );
+      const fieldMeta = objMetadata.fields.find(
+        field => field.name === item.data.soqlContext.fieldName
+      );
+      if (fieldMeta && !operatorsByType[fieldMeta.type].includes(item.label)) {
+        continue;
+      }
+    }
+
+    filteredItems.push(item);
+  }
+
+  return filteredItems;
+}
+
+async function expandPlaceholders(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const expandedItems = [...items];
+
+  for (const [index, item] of items.entries()) {
+    const parsedCommand = item.label.match(EXPANDABLE_ITEM_PATTERN);
+    if (parsedCommand) {
+      const commandName = parsedCommand[1];
+
+      const handler = expandFunctions[commandName];
+      if (!handler) {
+        console.log(`Unknown SOQL completion command ${commandName}!`);
```

Comment:
Not sure if this needs to be surfaced to the customer but we should at least track it with the telemetry service. Same for the rest of the `console.log`s in the rest of this file.

Reply:
Good point. I'll add telemetry

Reply:
Done